docs/proposals/subscriber-fan-in-fan-out.md

```diff
@@ -0,0 +1,189 @@
+# Switch to using Fan in Fan out / Observer pattern for subscribers
+
+This is not quite the same as a fan in fan out pattern but not a
+million miles away either.
+
+## Terminology
+
+**Observer:** suggestion to change the name from subscriber to observer
+as I think it more clearly conveys what they will be doing.
+For this doc I will continue to call them subscribers to avoid
+confusion.
+
+## Problem Description
+
+Our current subscriber pattern has some issues that have been outlined in the following
+[github issue](https://github.com/openshift/ansible-service-broker/issues/638)
+To surmise:
+
+- **Shared Channel with single subscriber**: There is a single buffered channel per topic (topics being broker
+actions such as provision, bind etc). All jobs of the given type write to
+this channel. This has the potential to cause a bottleneck, particularly
+as there is currently only a single subscriber reading from this channel and
+this subscriber does not read the next message until it has completed
+all of its work. Currently there is not too much happening in these
+subscribers, but if the work done in these subscribers were to increase
+or become computationally complex, it could begin to slow down throughput
+of messages from the jobs and potentially block them. The likely hood
+of this happening has been increased by the fact that we will soon allow
+fine grained status updates to sent by the APB developer via last operation
+updates.
+
+- **Separation of concerns**: While adding more subscribers would alleviate
+the previous issue, the current pattern means only one subscriber can
+deal with a given message at any one time. This makes it difficult to have a
+separation of concerns.  Ideally we would be able to have each subscriber
+responsible for doing just one thing (recording metrics adding various
+logs, persisting state etc all based on the msg it had been passed). This
+would make it simple to add more functionality based around these messages
+along with making testing simple and isolated.
+
+
+## Goal
+
+- Outline a design for passing messages, both user generated and system
+generated from executing APB jobs to each of a set of registered subscribers
+that removes the outlined problems and improves the overall design of the broker.
+
+## Subscriber design principal
+
+This proposal outlines changes to the subscribers and allowing
+the addition of more subscribers allowing them all to act on the same message.
+It is a key principal that the order these subscribers are called in
+should not matter, their work should be independent of any other subscribers work.
```

Comment:
Let's reword that last sentence, remove "it is a key principal that "

The order subscribers are called should not matter, their work should be independent of any other subscribers.